### DISCUSSION TITLE
Apicurio Registry integration with Kubernetes

### DISCUSSION THREAD

**Discussion by EricWittmann** [Apicurio/apicurio-registry#1498] [Ideas]
Discussion topic from `@eshepelyuk` on twitter:

> I see a several "operator" repositories here https://github.com/Apicurio?q=operator&type=&language=&sort=
> 
> Some are written in Go, but some of them are empty 
> * https://github.com/Apicurio/apicurio-studio-operator - empty but mentions Quarkus 
> * https://github.com/Apicurio/apicurio-registry-content-sync-operator - empty 
> 
> I am asking to better understand your intention and strategy for creating K8S native extensions for Apicurio, especially Registry.
> 
> I had in mind for sometime already the idea of creating the controller that can watch deployed resources ( Service, CRD) and populate Apicurio registry with API exposed by those resources, for instance poll OpenAPI and push them to Apicurio, thus creating the ultimate develoerp portal with all the documentaiton located in k8s cluster.
> 
> Maybe you have smth similar in mind and alrady working on that, If yes - please share your vision and tools you're going to use.
> 
> I'm interested in collaboratino and contribution in Java or Python.

**Comment by EricWittmann** [comment]
We do have a few operators currently.  Off the top of my head we have these:

* Apicurito installer operator (helm chart I think)
* Apicurio Registry installer operator (golang)
* Apicurio Studio installer operator (java/quarkus) **[in progress]**
* Apicurio Registry Sync operator (java/quarkus) **[in progress]**

I think the one that is most relevant to your discussion here is the Sync Operator:

https://github.com/Apicurio/apicurio-registry-content-sync-operator

The repository **was** empty but only because I hadn't yet merged the initial PR.  I've now merged that so you can have a look at what it does if you like.

@famartinrh has written that, and I think it is similar to what you are proposing.  It's not **quite** the same, assuming I understand your described use-case.  I think what you are suggesting is an operator that actually registers the OpenAPI design of actual running services deployed to Kubernetes.  So e.g. any micro-service that advertises its own API description via e.g. OpenAPI could be automatically registered in Apicurio Registry. 

For the record I think that's a fantastic idea.  And I think we could use the sync operator to make that happen.

**Reply by eshepelyuk** [reply]
Thanks for the explanation. I should take a look at `apicurio-registry-content-sync-operator`.
That's just amazing that you're not blindly writing k8s stuff in `golang` , 
but instead are open to more mainstream languages and frameworks, known by a wider audience of potential contributors.

**Reply by eshepelyuk** [reply]
Hello 

I've skimmed over the operator code and I consider that chosen strategy is more superior than my original proposal.
Having custom resources that describe schemas is extremely flexible, thus deployed *microservices* could provide as much as possible `Artifact` CRD to be stored in Apicurio Schema registry.

For instance if *microservice* installed with Helm, the chart could generate Avro `Artifact` for Kafka and OpenAPI `Artifact`  for REST API, picking up the content from application source code.

The only good addition would be to implement fetching OpenAPI from URL, so in case OpenAPI can't be generated in build time - the operator could fetch later OpenAPI declaration from the URL of running *microservice* instance.

One additonal note I'd like to raise - is the presentation of this operator. Since it not only a synchronization tool, but (at least for me) - it could transforms Apicurio Schema Registry into Develoler Portal, containing all the APIs used by the project \ platform.
Compare to those products
* https://docs.solo.io/gloo-portal/latest/
* https://www.getambassador.io/docs/edge-stack/latest/topics/using/dev-portal/

But unlikely beforementioned, Apicurio is not bound to any API Gatewa, could do more formats than OpenAPI and free !

**Reply by EricWittmann** [reply]
Great feedback thanks.  @famartinrh ^^^

I actually thought it had the ability to fetch the OpenAPI from a URL, but perhaps I'm mistaken.  

Hey @famartinrh perhaps a blog post about this operator and its capabilities makes sense!  :)

**Comment by eshepelyuk** [comment]
@EricWittmann @famartinrh I see that `apicurio-registry-content-sync-operator` doesn't provide Helm chart.
If you're interested I could help you with creating it.

**Reply by EricWittmann** [reply]
What would providing a Helm chart mean in this context?  I'm somewhat ignorant of Helm Chart use-cases and capabilities.

**Reply by eshepelyuk** [reply]
Hmm, 

How are you thinking the users are going to install the operator to cluster, via maven or kubectl ?
Helm charts is the best option to manage k8s native software installation to a cluster.

**Reply by famarting** [reply]
Currently we just provide the manifests so people would install the operator with `kubectl` , in the future we may also provide OLM, but I agree having a Helm chart is also pretty good.

Happy to receive some contributions on that :)

**Reply by eshepelyuk** [reply]
Sorry for delay, but I'm still interested in contributing Helm chart to this repo. 
Just didn't have time even to try the installation yet. 
Also, it would be great to have some instructions about the installation.

**Reply by eshepelyuk** [reply]
@famartinrh @EricWittmann is there a way to run this operator without maven and java ?
I can't find published docker images in this account https://hub.docker.com/u/apicurio.

**Reply by famarting** [reply]
Hi @eshepelyuk yes, it's possible to deploy this just with `kubectl` . I just updated the README file with some info, I hope it helps. https://github.com/Apicurio/apicurio-registry-content-sync-operator#how-to-deploy-this
The images for this project are only pushed to quay.io :)

**Reply by eshepelyuk** [reply]
Thanks for the readme. About the images, I hope you plan to release versioned ones sometimes in the future ?

**Reply by famarting** [reply]
yes sure, better if we stabilize it first and get some feedback and then we could create a 0.0.1 release. Or we could just tag the image with the snapshot version

**Reply by eshepelyuk** [reply]
Great, thank you.

**Reply by kmandal-volvo** [reply]
> @EricWittmann @famartinrh I see that `apicurio-registry-content-sync-operator` doesn't provide Helm chart.
> If you're interested I could help you with creating it.

Can you please helm to create helm chart for Apicurio -registry

**Reply by eshepelyuk** [reply]
I assume that eventually it should be the single chart. that could install both api-registry itself and sync-operator depending on passed values.

**Reply by kmandal-volvo** [reply]
> I assume that eventually it should be the single chart. that could install both api-registry itself and sync-operator depending on passed values.

I did not find any chart, please help to get the chart. As suggested I will manage in the value file

**Reply by eshepelyuk** [reply]
There's no any ready-to-use chart as far as I know.

**Reply by EricWittmann** [reply]
Correct - we don't have a helm chart for registry *or* for the sync operator.  I'm still a little confused as to why an operator needs a helm chart - can't operators be directly installed in kubernetes/openshift?

**Reply by eshepelyuk** [reply]
How directly installed ? With maven ? 
Helm is an universal tool to install anything into k8s (have no idea how openshift works and differs from k8s)
Operator is a just docker image running a "controller" that watches certain k8s resources and performs defined actions.

**Reply by famarting** [reply]
how? https://github.com/Apicurio/apicurio-registry-content-sync-operator#how-to-deploy-this this way for instance? `kubectl apply`

**Reply by kmandal-volvo** [reply]
I pull the latest image from dockerhub docker pull apicurio/apicurio-registry-operator:latest
Create a image push to our Azure kubernet registry .  Used https://raw.githubusercontent.com/apicurio/apicurio-registry-operator/master/docs/resources/install.yaml deployment descriptor but finally pod deployment failed ---- "error":"no matches for kind \"ApicurioRegistry\" in version \"apicur.io/v1alpha1\"","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/home/runner/go/pkg/mod/github.com/go-logr/zapr@v0.1.1/zapr.go:128\nmain.main\n\tapicurio-registry-operator/cmd/manager/main.go:144\nruntime.main\n\t/opt/hostedtoolcache/go/1.14.4/x64/src/runtime/proc.go:203"}

**Reply by kmandal-volvo** [reply]
> how? https://github.com/Apicurio/apicurio-registry-content-sync-operator#how-to-deploy-this this way for instance? `kubectl apply`

ComparisonError
failed to convert *unstructured.Unstructured to *v1beta1.Ingress: no kind "Ingress" is registered for version "networking.k8s.io/v1" in scheme "pkg/runtime/scheme.go:101"

**Reply by famarting** [reply]
those are errors with https://github.com/Apicurio/apicurio-registry-operator , if you could create an issue there

**Reply by eshepelyuk** [reply]
@kmandal-volvo 
please check https://github.com/Apicurio/apicurio-registry-content-sync-operator/issues/11
